// ng2-img-max/dist/src/ng2-img-max.module.js
import { NgModule } from "@angular/core";
import { Ng2ImgMaxService } from "./ng2-img-max.service";
import { ImgMaxSizeService } from "./img-max-size.service";
import { ImgMaxPXSizeService } from "./img-maxpx-size.service";
import { ImgExifService } from "./img-exif.service";
import { Ng2PicaModule } from "ng2-pica";
import * as ɵngcc0 from '@angular/core';
var Ng2ImgMaxModule = /** @class */ (function () {
    function Ng2ImgMaxModule() {
    }
Ng2ImgMaxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng2ImgMaxModule });
Ng2ImgMaxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Ng2ImgMaxModule_Factory(t) { return new (t || Ng2ImgMaxModule)(); }, providers: [
        { provide: ImgMaxPXSizeService, useClass: ImgMaxPXSizeService },
        { provide: ImgMaxSizeService, useClass: ImgMaxSizeService },
        { provide: ImgExifService, useClass: ImgExifService },
        { provide: Ng2ImgMaxService, useClass: Ng2ImgMaxService }
    ], imports: [[
            Ng2PicaModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2ImgMaxModule, { imports: function () { return [Ng2PicaModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ng2ImgMaxModule, [{
        type: NgModule,
        args: [{
                imports: [
                    Ng2PicaModule
                ],
                providers: [
                    { provide: ImgMaxPXSizeService, useClass: ImgMaxPXSizeService },
                    { provide: ImgMaxSizeService, useClass: ImgMaxSizeService },
                    { provide: ImgExifService, useClass: ImgExifService },
                    { provide: Ng2ImgMaxService, useClass: Ng2ImgMaxService }
                ]
            }]
    }], function () { return []; }, null); })();
    return Ng2ImgMaxModule;
}());
export { Ng2ImgMaxModule };
