// ng2-daterangepicker/esm2015/lib/ng2-daterangepicker.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { DaterangepickerComponent } from './ng2-daterangepicker.component';
import * as ɵngcc0 from '@angular/core';
let Daterangepicker = class Daterangepicker {
};
Daterangepicker.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Daterangepicker });
Daterangepicker.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Daterangepicker_Factory(t) { return new (t || Daterangepicker)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Daterangepicker, { declarations: function () { return [DaterangepickerComponent]; }, exports: function () { return [DaterangepickerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Daterangepicker, [{
        type: NgModule,
        args: [{
                declarations: [DaterangepickerComponent],
                imports: [],
                exports: [DaterangepickerComponent]
            }]
    }], null, null); })();
export { Daterangepicker };
